### ISSUE TITLE
'\g' incorrectly flagged as invalid escape sequence

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#657]
In GitLab by @johnthagen on Nov 30, 2018, 19:24

*Please describe how you installed Flake8*

```
$ python3 -m venv venv
...
(venv) $ pip install flake8
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
$ flake8 --bug-report
{
  "dependencies": [
    {
      "dependency": "setuptools",
      "version": "39.1.0"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.7.0",
    "system": "Darwin"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.4.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.0.0"
    }
  ],
  "version": "3.6.0"
}
```

*Please describe the problem or feature*

This code run correctly:

```py
import re

s = re.sub(pattern='(?P<type>Apple|Banana)',
           repl='@{\g<type>}',
           string='Apple')

print(s)
```

And prints: `@{Apple}`

But running `flake8`:

```
mod.py:3:4: W605 invalid escape sequence '\g'
```

But `\g` is a valid string for use in the `re` library, and runs correctly:

https://docs.python.org/3/library/re.html#regular-expression-syntax

**Comment by asottile** [comment]
In GitLab by @asottile on Nov 30, 2018, 19:31

A couple things -- first `flake8`'s rules are not provided by `flake8` but by the tools it wraps -- in this case this error comes from [`pycodestyle`](https://github.com/pycqa/pycodestyle).

`\g` is in fact an invalid escape sequence.  In python3.8+ it will be a syntax error (you can enable it earlier by escalating warnings to errors):

```console
$ python3.7 -Werror -c '"\g"'
  File "<string>", line 1
SyntaxError: invalid escape sequence \g
```

`pycodestyle` chooses to helpfully flag this for you such that you can fix it before python3.8 rolls around.

You should use a raw string instead, or manually escape the backslash:

```python
r'@{\g<type>}'
'@{\\g<type>}'
```

If you're looking for a way to automatically fix these for you, I've written a tool called [pyupgrade](https://github.com/asottile/pyupgrade) which has a specific part which [fixes invalid escape sequences](https://github.com/asottile/pyupgrade#invalid-escape-sequences)

**Comment by asottile** [comment]
In GitLab by @asottile on Nov 30, 2018, 19:32

notice also that even the gitlab syntax highlighting called out your `\` above :)

**Comment by asottile** [comment]
In GitLab by @asottile on Nov 30, 2018, 19:32

closed

**Comment by asottile** [comment]
In GitLab by @johnthagen on Nov 30, 2018, 19:35

@asottile Thanks for the detailed explanation! I wish Python had these kinds of warnings on by default, but I guess that's why we have tools like `flake8`.

In my case I definitely want a raw string, and am glad this was caught before 3.8 comes out.

Thanks! :thumbsup:

**Comment by asottile** [comment]
In GitLab by @asottile on Nov 30, 2018, 19:39

There is a warning in python3.6 but it's hidden by default (sadly) -- they fixed this in python3.7 to show [deprecation warnings in code invoked by `__main__`](https://docs.python.org/3/library/warnings.html#default-warning-filter)
